Added Level-3 support for deployment & miscs

N0OBSTUDENT opened the pull request:
-The real race environment could read the actual track layout and "randomize" the layout from MoCapas nominal track info
-Refactored checks.py, making it independent to `ROSConnector`. The only file that calls it will be real_race_env.py.
-Reorganized real race environment reset procedure. Track objects are checked before the drone connects. Now the students do not need to wait a long time for drone connection before they  know whether they are placing the track objects at the correct position
-Now with `--save_track some_config_name.toml` option while calling `python deploy.py`, the actual track layout with other the configuration from the original config file will be stored as backup in `config/some_config_name.toml`, so that the students could check setup after the real-world run.

N0OBSTUDENT committed:
Level 3 deployment

-Added the track.randomize config readibility to real race environment. The real race environment could read the actual track layout and "randomize" the layout as nominal track info
-Refactored checks, making it independent to ROSConnector
-Reorganized real race environment reset procedure. Track objects are checked before the drone connects. Now the students do not need to wait a long time for drone connection before they  know they are not placing the track objects in the correct position

N0OBSTUDENT committed:
Reorganized code in real race environment

N0OBSTUDENT committed:
Tested and fixed minor bugs to make it works.

N0OBSTUDENT committed:
Passed ruff check

N0OBSTUDENT committed:
Added save track function.

-Now with "--save_track *.toml" option while deploying, the actual track layout overwriting the original configuration file will be written in config/

amacati commented:
@Rather1337 From a first glance, some files seem to lack ruff formatting. Can we run ruff format in the CI and check if if changed stuff? I think there is a "display changes only" flag that we can use to catch this in PRs.

ratheron commented:
I says ruff passed in the checks. What do you mean exactly

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -114,52 +120,16 @@ def __init__(
-        if options.get("check_drone_start_pos", True):
-            pos = self.drones.pos[self.rank, ...]
-            check_drone_start_pos(pos, self.randomizations, self.drone_name)
+        self._prepare_track(deploy_config = options.deploy, track = options.env.track)
```
The name "prepare" is pretty undescriptive. I would not be sure what this method does just by looking at the name. Can we make these clearer?

N0OBSTUDENT replied:
Do you have any suggestion on this? It involves updating the actual gate pose, obstacle position from ROSConnector and checks the track. I can't think of a name that summarize these steps.

amacati replied:
Then maybe it should be two steps :D `self._update_track_poses` and `self._check_race_track` seem pretty descriptive

N0OBSTUDENT replied:
OKay, then I do not even need the query track function. It's executed once. I separated everything just to make functions not too long.

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -265,6 +235,31 @@ def info(self) -> dict:
         """Return an info dictionary containing additional information about the environment."""
         return {}
 
+    def _query_track_obj_pose(self) -> tuple[NDArray, NDArray, NDArray]:
```
I am not a huge fan of this method, since it just bundles stuff together that we already should have set up anyways, and its signature is a bit confusing. We say pose, but only return the pos for obstacles? We say obj, but you have to check the signature to know it's gate pos, gate quat, obstacle pos instead of gate pose, obstacle pose, drone pose, etc.

N0OBSTUDENT replied:
We need that info to update the pose of the gates and obstacles  and check the track. Originally it was written in check.py and I tried to decouple ROSConnector with check functions.
What about returning a ConfigDict that includes these info? That will make this clearer.

amacati replied:
Do we even need to wrap it in a separate function?

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -265,6 +235,31 @@ def info(self) -> dict:
+            for i in range(self.n_obstacles):
+                obstacle_pos[i, ...] = pos[f"obstacle{i + 1}"]
+        except KeyError as e:
+            logger.error(f"Could not find track objects in the ROS TF tree: {e}. Have you enabled the track objects in Vicon/ started the motion capture tracking node?")
```
This line blows the 100 chars limit by a fair amount.

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
             # Make sure to use the legacy estimator with the state interface
 
-    def _connect_to_drone(self, radio_id: int, radio_channel: int, drone_id: int) -> Crazyflie:
+    def _prepare_track(self, deploy_config: ConfigDict, track: ConfigDict):
```
See comment on naming further up.

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
 
-    def _connect_to_drone(self, radio_id: int, radio_channel: int, drone_id: int) -> Crazyflie:
+    def _prepare_track(self, deploy_config: ConfigDict, track: ConfigDict):
+        assert (not track.get("randomize", False)) or (track.get("randomize", False) and deploy_config.get("real_track_objects", True)), "With a randomly generated track, the deploy.real_track_objects must be set to true."
```
Checks beyond sanity asserts should be real errors, not asserts. This is not yet consistent across the code, but we should stick to it for new code.

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
+                        gates=self.gates,
+                        obstacles=self.obstacles,
+                        rng_config=self.randomizations,
+                )
```
It should be possible to clean this block up to be tidier. As it is, it's hard to read and reason about
```suggestion
        if not deploy_config.get("real_track_objects"):
            return
        self.gates.pos, self.gates.quat, self.obstacles.pos = self._query_track_obj_pose()
        if track.randomize:
            self.gates.nominal_pos, self.gates.nominal_quat, self.obstacles.nominal_pos = randomize_track(
                gates_pos=self.gates.pos,
                gates_quat=self.gates.quat,
                obstacles_pos=self.obstacles.pos,
                rng_config=self.randomizations
            )
            # TODO: The gate safety limits are now hardcoded: 0.5 meters inside the drone safety limits.
            limit_low = np.array(track.safety_limits.pos_limit_low[:2]) + 0.5
            limit_high = np.array(track.safety_limits.pos_limit_high[:2]) - 0.5
            check_gates_layout(gates=self.gates, limits={'pos_limit_low' : limit_low, 'pos_limit_high' : limit_high})
        if deploy_config.get("check_race_track", True):
                check_race_track(
                    gates=self.gates,
                    obstacles=self.obstacles,
                    rng_config=self.randomizations,
            )
```

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
+                        rng_config=self.randomizations,
+                )
+    
+    def _prepare_drones(self, deploy_config: ConfigDict, track: ConfigDict):
```
Naming comment applies

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
+    
+    def _prepare_drones(self, deploy_config: ConfigDict, track: ConfigDict):
+        if track.randomize:
+            # The starting position drones.pos has must be set here even though they are not used in _step(). Since they are the target pose for return controller.
```
This is way >100 lines. @Rather1337 can a HiWi double-check how this passes ruff?

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -362,6 +423,22 @@ def _apply_drone_settings(self):
+        )
+        try:  # Make sure to close the connection if anything goes wrong
+            pos = ros_connector.pos[self.drone_name]
+            
```
```suggestion
```

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -362,6 +423,22 @@ def _apply_drone_settings(self):
         self.drone.param.set_value("flightmode.stabModeYaw", 1)
         time.sleep(0.1)  # Wait for settings to be applied
 
+    def _query_drone_pose(self) -> tuple[NDArray, NDArray]:
```
I don't think this is a good idea. We have a connector running anyways that should track the drone. Why are we creating more? They are heavy objects with substantial init times.

mschuck-rai replied:
It's just a wrapping for 
`pos, quat = self._ros_connector.pos[self.drone_name], self._ros_connector.quat[self.drone_name]` ?

N0OBSTUDENT replied:
The ROSConnector was also created twice in the original code, but the first time was in check.py . I thought that's done intentionally so I copied the procedure here.

In my code, the `self._ros_connector` will not be initialized until gate, obstacle and drone pose checks are passed. So I made a temporary connector here which does not send commands but only query pose of the drone. But I could certainly initialize `self._ros_connector` at the first place.

amacati replied:
I'm not sure why it was created twice, but we should try to reuse them as much as possible without becoming unergonomic.

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
+            if deploy_config.get("check_drone_start_pos", True):
+                real_pos, real_quat = self._query_drone_pose()
+                nominal_pos = np.array(self.drones.pos[self.rank])
+                check_drone_start_pos(nominal_pos=nominal_pos, real_pos=real_pos, rng_config=self.randomizations, drone_name=self.drone_name)
```
I'm happy to sacrifice a few ms for the sanity check. Also, it's nice to reduce logic complexity.

```suggestion
        if deploy_config.get("check_drone_start_pos", True):
            real_pos, real_quat = self._query_drone_pose()
            nominal_pos = np.array(self.drones.pos[self.rank])
            check_drone_start_pos(nominal_pos=nominal_pos, real_pos=real_pos, rng_config=self.randomizations, drone_name=self.drone_name)
```

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,7 +284,73 @@ def send_action(self, action: NDArray):
+            radio_channel=self.drone_ch, drone_id=self.drone_id
+        )
+
+    def _establish_drone_connection(self, radio_channel: int, drone_id: int):
```
Please keep namings as simple as possible. Wanna connect to the drone? -> `connect_to_drone`

```suggestion
    def _connect_to_drone(self, radio_channel: int, drone_id: int):
```

mschuck-rai commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -114,52 +120,16 @@ def __init__(
             n_drones=self.n_drones, n_gates=self.n_gates, n_obstacles=self.n_obstacles
         )
-        # Establish drone connection
-        self._drone_healthy = mp.Event()
```
Please do not remove attributes that are set outside of `__init__`. Every self attribute that is set at some point should already be initialized here. Also, why did you remove the `ROSConnector`?

N0OBSTUDENT replied:
That's because I could not think of a better way to work around the problem here:

Any race track check failure in `_reset()` will throw an Exception. The Exception will trigger the try-catch-finally in deploy.py, which executes  `return_to_start()` and codes killing the drone. However, the drone is planned be connected at the very last step of `_reset()`, so obviously there is no connection to the drone by that time, which will throw two extremely long exceptions, making the console display a giant stack of Exceptions.

My current work-around is just setting them as `None` upon initialization of the class and initialize them in `_reset()`, so I could have a simple if-else in `self.close()` to judge whether to trigger the return controller and kill the drone.

The other work-around I thought of at the very beginning is that we could do all stuffs except drone connection in `__init__`, since there is certainly no `env.close` called by then. But the problem is that, track checking requires config.deploy field, which is currently not an argument to `__init__` . I suppose you are not willing to add more arguments, so I did not do that.


Edit: Maybe I just need to move the track checking  and drone position checking out of the class, and execute them before initializing the Environment object. Then there will not be a god class there.

Do you have any advice on that?

amacati replied:
I think both is reasonable. It's fine that the environment asserts everything is as expected before starting, but we can also pull that out. Regarding the return to start, it should be feasible to check if the drone has taken off or not, right? At worst, this can be a flag that is set or determined by checking if the drone is close to the ground and moving.

mschuck-rai commented on lsy_drone_racing/utils/utils.py:
```diff
@@ -79,6 +80,60 @@ def load_config(path: Path) -> ConfigDict:
+            'env' : config.env.to_dict(),
+            'sim' : config.sim.to_dict()
+    }
+    # Now the randomized generation should be set to false
```
Please do not comment what the code does. This is just a repeat of the line below. Comment on **why** you are doing this.

mschuck-rai commented on lsy_drone_racing/utils/utils.py:
```diff
@@ -79,6 +80,60 @@ def load_config(path: Path) -> ConfigDict:
+    ]
+    with open(output_path, "w") as f:
+        toml.dump(dict_to_store, f)
+
```
This is _very_ hard to read. Please refactor before merging. Why was this function necessary in the first place?

N0OBSTUDENT replied:
To save the current layout of the track.
In level-3  there are "completely randomly generated" tracks and in deployment stage the track is defined by the placement of the gates and obstacles in real-world, i.e. not from the config toml file. So the students may want to rerun the case in simulation but they don't have the track layout. The function could backup the real-world setting for them to rerun in sim.

N0OBSTUDENT replied:
But I will try to make it look nicer.

amacati replied:
I like the idea

N0OBSTUDENT replied:
Would it be a good practice to have this in the `RealRaceCoreEnv`? I could write something like `get_level_config()->Dict[str, Any]` in `RealRaceCoreEnv`, which just outputs the dict of updated configuration and we could dump that dict later in deploy.py. Then I do not need  so many arguments and those ugly unwrapping operations.

mschuck-rai commented on scripts/deploy.py:
```diff
@@ -17,24 +17,27 @@
     rclpy.init()
     config = load_config(Path(__file__).parents[1] / "config" / config)
+    # Overwrite the config file controller
+    config.controller.file = controller or config.controller.file
```
Do not use `or` in case of `None`. Use explicit `is None` checks.

N0OBSTUDENT replied:
Will fix that. I thought you thought it was a good practice since it was there in the original one.

amacati replied:
No, it's a bad habit of mine, and I try to get rid off it because it looks so concise, but the outcomes are different. E.g. `1 if y is None else y` with `y = ""` will yield `""`, but `y or 1` will yield `1`.

mschuck-rai commented on scripts/deploy.py:
```diff
@@ -44,17 +47,19 @@ def main(config: str = "level2.toml", controller: str | None = None):
         sensor_range=config.env.sensor_range,
         control_mode=config.env.control_mode,
     )
+    
```
```suggestion
```

mschuck-rai commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -15,63 +14,101 @@
+
+    low, high = rng_config.obstacle_pos.kwargs.minval, rng_config.obstacle_pos.kwargs.maxval
+    for i, (pos, nominal_pos) in enumerate(zip(obstacles.pos, obstacles.nominal_pos)):
+        check_bounds(f"obstacle{i + 1}", np.array(pos[:2]), np.array(nominal_pos[:2]), np.array(low[:2]), np.array(high[:2]))
```
Do we need all the np casts? They make the code hard to read

N0OBSTUDENT replied:
Hmm no. I did that just to comply to the function signature. I could make argument of `check_bounds()` be `ArrayLike` and cast them in that.

mschuck-rai reviewed (changes requested):
A couple of comments on the PR. It would be great if you could confirm that the options (with/without track checks etc.) have been tested in real.

mschuck-rai commented:
> I says ruff passed in the checks. What do you mean exactly

The ruff CI checks with `ruff check`, but `ruff format` would make changes to the files. We should check that PRs have run the latter. E.g. there are instances of two line breaks between class methods etc. We should catch this.

mschuck-rai commented:
Thanks for working on this btw @N0OBSTUDENT

N0OBSTUDENT commented:
> A couple of comments on the PR. It would be great if you could confirm that the options (with/without track checks etc.) have been tested in real.

I have tested all cases I could think of  (w or w/o real track objects, w or w/o track check, w or w/o fully randomly generated tracks) in the lab, but I will test them and commit again after reorganizing the codes according to the comments. Could I further add commits upon this PR? Or I can only re-open one with new commits?

amacati commented:
> Could I further add commits upon this PR? Or I can only re-open one with new commits?

Just keep pushing to your branch, it will get auto-updated here

N0OBSTUDENT committed:
Reorganized code-not tested and not finished
- Now the environment code is shorter.
- TODO: move the track check procedure out of the environment class, so that the class definition will be shorter

N0OBSTUDENT committed:
Reorganize procedures-still not tested

N0OBSTUDENT committed:
Merge branch 'utiasDSL:dev' into dev

N0OBSTUDENT committed:
Added take off flag

N0OBSTUDENT committed:
Added docstring for check_bounds

N0OBSTUDENT committed:
Edited docstring for update_level_config

N0OBSTUDENT committed:
Merge branch 'dev' of https://github.com/N0OBSTUDENT/lsy_drone_racing into dev

N0OBSTUDENT committed:
Refactored rotation check: now rotations are checked in three axes.
Fixed minor issue: ROSConnector timeout values are increased to 10 seconds.
Fixed minor issue: The update_level_config function missed a key in dict query, now it works.
Added take off flag: A flag will stay True in RealRaceCoreEnv after taking off(once z coordinate > 0.2 meter), and will be queried when calling env.close(). If the drone has not "taken off", the return back controller will not be called.
Tested cases: See https://docs.google.com/spreadsheets/d/1a_WvakzvkbI9FE5ZeVwlMI_acAK3nL8hTaW2EinpsPs/edit?usp=sharing

N0OBSTUDENT committed:
Added docstring to _check_drones in RealRaceCoreEnv

N0OBSTUDENT commented:
@amacati Hi, I have modified the code accordingly. Please take a look and see whether it needs any further improvements.
For test cases please refer to [this](https://docs.google.com/spreadsheets/d/1a_WvakzvkbI9FE5ZeVwlMI_acAK3nL8hTaW2EinpsPs/edit?usp=sharing)

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -102,65 +107,40 @@ def __init__(
         self.drone_names = [f"cf{drone['id']}" for drone in drones]
         self.drone_name = self.drone_names[rank]
-        self.channel = drones[rank]["channel"]
+        self.drone_ch = drones[rank]["channel"]
```
Please leave this as drone channel. I would have to guess what `self.drone_ch` is, but `self.drone_channel` is very clear and not overly verbose.

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -102,65 +107,40 @@ def __init__(
-        )
-        self._last_drone_pos_update = 0  # Last time a position was sent to the drone estimator
-
+        self._drone_healthy.clear()
```
Why is the drone being set to not healthy? I do not see an error state

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -208,11 +188,13 @@ def obs(self) -> dict[str, NDArray]:
+            np.float32
+        )
+        mask_obs = self.data.obstacles_visited[..., None]
+        obstacles_pos = np.where(mask_obs, self.obstacles.pos, self.obstacles.nominal_pos).astype(
```
Was there a reason for this change? Was it not clear what the code was doing?

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+                    obstacles_pos=self.obstacles.pos,
+                    rng_config=self.randomizations,
+                )
+            )
```
I am not a fan of the mixing in here and in `_check_track_layout`. This function gets long and does two things depending on the settings. Also, if the function starts with if ...: and skips everything if that condition is not set, it's best to do the check before calling the function and then do the stuff that the function promises.

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+                logger.error(
+                    f"Could not find track objects in the ROS TF tree: {e}. \
+                             Have you enabled the track objects in Vicon/ \
+                             started the motion capture tracking node?"
```
```suggestion
                    f"Could not find track objects in the ROS TF tree: {e}. Have you enabled the "
                    "track objects in Vicon and started the motion capture tracking node?"
```

amacati replied:
It's also not clear why you would want to catch, log, and then reraise the error. Were students unsure what was happening?
In general, try to use as few try-except blocks as possible. If something fails early that's a good thing. There are exceptions of course, e.g. the try finally around the connector to prevent processes hanging indefinitely

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
-    def _connect_to_drone(self, radio_id: int, radio_channel: int, drone_id: int) -> Crazyflie:
+    def _update_track_poses(self, deploy_config: ConfigDict, track: ConfigDict):
+        """Update the track poses from the motion capture system."""
+        if deploy_config.get("real_track_objects", True):
```
This should be checked before the function is called

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+        if deploy_config.get("check_race_track", True):
+            check_race_track(
+                gates=self.gates, obstacles=self.obstacles, rng_config=self.randomizations
+            )
```
Same thing with the mixing and the checking if we actually do something. The functions themselves make a lot of sense and also make the code in reset much nicer, but this mixing should be sorted out

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+            drone_pos, drone_quat = (
+                self._ros_connector.pos[self.drone_name],
+                self._ros_connector.quat[self.drone_name],
+            )
```
Avoid multi-line statements whenever possible. It makes the code so much more readable

```suggestion
            drone_pos = self._ros_connector.pos[self.drone_name]
            drone_quat = self._ros_connector.quat[self.drone_name]
```

N0OBSTUDENT replied:
Okay, I suppose it was a existing practice

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+            real_pos, _ = (
+                self._ros_connector.pos[self.drone_name],
+                self._ros_connector.quat[self.drone_name],
+            )
```
...
```suggestion
            real_pos = self._ros_connector.pos[self.drone_name]
```

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+            )
+            drone_pos[2] = 0.01  # Set z to ground level
+            drone_rpy = R.from_quat(drone_quat).as_euler("xyz", degrees=False)
+            self.drones.pos[self.rank], self.drones.rpy[self.rank] = drone_pos, drone_rpy
```
Not sure why this is in here. We are checking if the drone is ready to start. Overwriting the position should not be mixed with the check. Same as for the other functions.

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -390,8 +468,6 @@ def wait_for_action(dt: float):
         pos = self._ros_connector.pos[self.drone_name]
         vel = self._ros_connector.vel[self.drone_name]
         # This quick check prevents us from engaging the return controller if we havent even started yet.
```
This comment does not make sense anymore

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -418,6 +494,43 @@ def _jit(self):
+            }
+            for i in range(self.drones.pos.shape[0])
+        ]
+        return ConfigDict(result_config)
```
Does this not work? Again, we want as less indentation, brackets etc as possible to improve readability.
```suggestion
        config = config.copy()
        # We store the real-world track layout, so randomization must be disabled
        config.env.track.randomize = False
        # Overwrite the original positions and orientations with the measured ones
        for i in range(self.gates.pos.shape[0]):
            config.env.track.gates.pos[i] = self.gates.pos[i].tolist()
            config.env.track.gates.rpy[i] = R.from_quat(self.gates.quat[i]).as_euler("xyz").tolist()
        for i in range(self.obstacles.pos.shape[0]):
            config.env.track.obstacles.pos[i] = self.obstacles.pos[i].tolist()
        for i in range(self.drones.pos.shape[0]):
            config.env.track.drones.pos[i] = self.drones.pos[i].tolist()
            config.env.track.drones.rpy[i] = (
                R.from_quat(self.drones.quat[i]).as_euler("xyz").tolist()
            )
        return ConfigDict(config)
```

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -3,10 +3,9 @@
 
 import logging
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Union
```
Do not use Unions. Use pipe syntax instead.

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -16,66 +15,125 @@
 
 
-def check_race_track(
+def randomize_track(
```
The file is called `checks.py`. This is not a check.

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -16,66 +15,125 @@
+    return randomized_gate_pos, randomized_gate_quat, randomized_obstacle_pos
+
+
+def check_gates_in_bound(gates: ConfigDict, limits: ConfigDict):
```
Please do not use `ConfigDict`s everywhere. It gets hard to understand what the function does. If you want to check positions, there is no need to pass around dictionaries with nested data. Just pass in the plane data as a `NDArray`.

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -16,66 +15,125 @@
+            )
+
+
+def check_race_track(gates: ConfigDict, obstacles: ConfigDict, rng_config: ConfigDict):
```
Same here

N0OBSTUDENT replied:
But here if I pass the nominal gate positions, gate rotations, obstacle positions, randomization upper/lower bound of gate positions, rotations, obstacle positions, that makes the signature very long. What should be `ConfigDict` and what should not?

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -16,66 +15,125 @@
 
 
 def check_bounds(name: str, actual: NDArray, desired: NDArray, low: NDArray, high: NDArray):
-    """Check if the actual value is within the specified bounds of the desired value."""
+    """Check if the actual value is within the specified bounds of the desired value.
+
+    Raise error when the actual value exceeds the low or high boundaries.
+
+    Args:
+        name: Name of the object being checked.
+        actual: NDArray of real value and is expected to be [N, ].
+        desired: NDArray of nominal value and is expected to be [N, ].
+        low: NDArray of the minimum value of (actual - desired) and is expected to be [N, ]
+        high: NDArray of the maximum value of (actual - desired) and is expected to be [N, ]
+    """
```
Nothing in this function assumes a shape. Also, do not repeat types in the doc string. They are contained in the function signature already.

```suggestion
    """Check if the actual value is within the specified bounds of the desired value.

    Args:
        name: Name of the object being checked.
        actual: Values to check.
        desired: Reference values.
        low: Lower bound. Minimum permissible value of (actual - desired).
        high: Upper bound. Maximum value of (actual - desired).
    
    Raises:
            RuntimeError: The values are not in the permissible interval.
    """
```

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -86,11 +144,46 @@ def check_bounds(name: str, actual: NDArray, desired: NDArray, low: NDArray, hig
+    actual_rot: R,
+    desired_rot: R,
+    low: Union[np.ndarray, float],
+    high: Union[np.ndarray, float],
```
```suggestion
    low: NDArray | float,
    high: NDArray | float,
```

amacati commented on lsy_drone_racing/utils/checks.py:
```diff
@@ -86,11 +144,46 @@ def check_bounds(name: str, actual: NDArray, desired: NDArray, low: NDArray, hig
+            raise RuntimeError(
+                f"{name} exceeds higher rotation tolerances ({high}).\n"
+                f"Rotation is: {actual}, should be: {desired}"
+            )
```
This needs to be revised. We cannot change the metric of angular measurement and with it the whole behaviour of the function depending on whether you pass in a 3D array or a float. This is error-prone and confusing at best.

N0OBSTUDENT replied:
That's due to legacy concerns. Shall I just keep the 3D array? The scalar is actually not used now

amacati commented on scripts/deploy.py:
```diff
@@ -25,16 +26,20 @@
         config: Path to the competition configuration. Assumes the file is in `config/`.
         controller: The name of the controller file in `lsy_drone_racing/control/` or None. If None,
             the controller specified in the config file is used.
+        save_track: Path to store the current competition configuration. The file is stored in `config/`. If None, the configuration will not be saved.
```
`save_track` sounds like a flag, not a path.

amacati commented on scripts/deploy.py:
```diff
@@ -25,16 +26,20 @@
     rclpy.init()
     config = load_config(Path(__file__).parents[1] / "config" / config)
+    if controller is not None:
+        config.controller.file = controller
```
config.controller.file should be a `Path` if possible

N0OBSTUDENT replied:
The `config.controller.file` is originally a string loaded from the toml file
``` 
[controller]
file = "state_controller.py"
```
And I do not think that the type should be changed. It has to be `str` in order to get concatenated afterwards
```  
control_path = Path(__file__).parents[1] / "lsy_drone_racing/control"
controller_path = control_path / config.controller.file
```

amacati commented on scripts/deploy.py:
```diff
@@ -44,17 +49,21 @@ def main(config: str = "level2.toml", controller: str | None = None):
-            "real_track_objects": config.deploy.real_track_objects,
-        }
-        obs, info = env.reset(options=options)
+        obs, info = env.reset(options=config)
```
I am not a fan of passing the complete config as options. We do not want god objects that contain everything, and if possible we should also limit the use of god config objects. Why was this changed?

N0OBSTUDENT replied:
Now to check whether the track is in place, both the field `track.randomize` and `deploy` are needed. This is because the `track.randomize` field gives out whether the track is randomly generated, and this is necessary for checking procedures.

The most important thing is that I need to check whether the `deploy.real_track_object` is `True` when `track.randomize` is set to `True`. If that's skippable, I can do everything regarding randomly generated track in `__init__` and get rid of `deploy` field.

I can not think of any work around, except changing the toml file structure.

amacati commented on scripts/deploy.py:
```diff
@@ -44,17 +49,21 @@ def main(config: str = "level2.toml", controller: str | None = None):
+            output_path = Path(__file__).parents[1] / "config" / save_track
+            assert output_path.suffix == ".toml", (
+                f"Configuration file has to be a TOML file: {output_path}"
+            )
```
```suggestion
            if not output_path.suffix == ".toml":
                raise ValueError(f"Configuration file has to be a TOML file: {output_path}")
```

amacati replied:
Prefer real errors over asserts if the error is a likely possibility. Asserts should not be part of the intended control flow. They are meant to sanity check stuff. In fact, you can turn off all asserts with `python -O ...`.

amacati reviewed (commented):
Looks much better, but there are multiple things that still need to be addressed.

amacati commented:
Please also resolve any comments I made that you fixed or addressed to keep the conversation clean.

The real-world test sheet looks amazing. Thanks a lot for going through all of the options.

amacati commented on lsy_drone_racing/envs/real_race_env.py:
```diff
@@ -289,36 +271,130 @@ def send_action(self, action: NDArray):
+            Otherwise, the function will check whether the track is within the tolerance limits.
+        """
+        if track.randomize:
+            # TODO: The gate safety limits are now hardcoded: 0.5 meters inside the drone safety limits.
```
This should not be hard coded :D

N0OBSTUDENT committed:
Update lsy_drone_racing/utils/checks.py

Co-authored-by: Martin Schuck <57562633+amacati@users.noreply.github.com>

N0OBSTUDENT committed:
Rename variables and update drone health check logic.

N0OBSTUDENT committed:
Update lsy_drone_racing/utils/checks.py

Co-authored-by: Martin Schuck <57562633+amacati@users.noreply.github.com>

N0OBSTUDENT committed:
Refactored update_level_config

Co-authored-by: Martin Schuck <57562633+amacati@users.noreply.github.com>

N0OBSTUDENT committed:
Moved randomize_track function from utils.checks to envs.utils
The _reset function now only accept deploy configuration instead of the whole configuration. Correspondingly, a _predefined_track boolean is stored to enable different track check behaviors.
Moved condition checks to _reset function, making it more explicit.
The check_rotation function in checks.py now only accepts NDArray as input.

N0OBSTUDENT committed:
Merge branch 'dev' of https://github.com/N0OBSTUDENT/lsy_drone_racing into dev

N0OBSTUDENT committed:
Merge branch 'dev' into dev

N0OBSTUDENT committed:
Dealt with ruff checks: lines were too long

N0OBSTUDENT committed:
Merge branch 'dev' of https://github.com/N0OBSTUDENT/lsy_drone_racing into dev

N0OBSTUDENT committed:
Format files to pass ruff checks

N0OBSTUDENT committed:
Sync lock files

amacati committed:
Merge branch 'dev' into fix-pr

amacati committed:
Remove track randomization from deploy. Move stateful attributes into EnvState.

N0OBSTUDENT committed:
Kick the generated track feature out of deploy and RealRaceCoreEnv

N0OBSTUDENT committed:
Add save_track_as_config.py
To save a track layout, call scripts/save_track_as_config.py --config some_config_file.toml --save_config_to output_config_file.toml

N0OBSTUDENT committed:
Merge branch 'dev' of https://github.com/N0OBSTUDENT/lsy_drone_racing into dev
Fixed "EnvData still needs a drone_connected attribute, I forgot to add that" by Martin

N0OBSTUDENT committed:
Fixed naming inconsistency

N0OBSTUDENT committed:
Fixed out-of-date check_track.py and moved query_drone_poses and query_track_poses into envs/utils.py

N0OBSTUDENT committed:
Move all functions using drone_estimators out of envs/utils.py since in tests environment it is not installed and therefore import errors will be raised.

N0OBSTUDENT committed:
Make self.data.reset called before connecting the drone
Switch function defination sequence

amacati committed:
Refactor track helpers

amacati committed:
Remove unnecessary dicts in check_track

amacati committed:
Add warning and TODO docs to check_rotation

amacati committed:
Fix typo

amacati committed:
Fix CI
